synapse/storage/prepare_database.py

```diff
@@ -133,22 +133,20 @@ def prepare_database(
 
             # if it's a worker app, refuse to upgrade the database, to avoid multiple
             # workers doing it at once.
-            if (
-                config.worker.worker_app is not None
-                and version_info.current_version != SCHEMA_VERSION
-            ):
+            if config.worker.worker_app is None:
+                _upgrade_existing_database(
+                    cur,
+                    version_info,
+                    database_engine,
+                    config,
+                    databases=databases,
+                )
+            elif version_info.current_version < SCHEMA_VERSION:
```

Comment:
The change from `!=` to `<` is the real fix here, I believe?

Is it worth adding a comment saying "If the database hasn't been updated yet, refuse to start" or something?

Reply:
that's pretty much what the comment at line 134 is saying, though I kindof agree that it might be clearer to move it down now.